tests/test_workflows.py

```diff
@@ -124,4 +133,85 @@ async def event_workflow(c: Context):
         assert s.w_run_id in related_runs
 
 
-# TODO: test concurrent
+@pytest.fixture
+def utils_state():
+    return {"current_time": None, "rand_value": None, "unique_id": None}
+
+
+@async_test
+async def test_utils_methods(t: WorkflowTest, utils_state):
+    @workflow()
+    async def utils_workflow(context: Context):
+        a = await context.utils.now()
+        b = await context.utils.random(1, 10)
+        c = await context.utils.uuid4()
+        x = await context.save(lambda: "2024-11-29T13:52:08.859245")
+        if (
+            utils_state["current_time"] is None
+            and utils_state["rand_value"] is None
+            and utils_state["unique_id"] is None
+        ):
+            utils_state["current_time"] = a
+            utils_state["rand_value"] = b
+            utils_state["unique_id"] = c
+        else:
+            assert a == utils_state["current_time"]
+            assert b == utils_state["rand_value"]
+            assert c == utils_state["unique_id"]
+            assert isinstance(a, datetime)
+            assert isinstance(x, str)
+
+        await context.sleep(timedelta(seconds=1))
+        return a, b, c, x
+
+    backends = [
+        FSBackend("./logs"),
+        RedisBackend(host="localhost", port=6380, password="password"),
+    ]
+
+    for backend in backends:
+        s = await t.step(backend).exec_workflow(utils_workflow)
+        now, rand, uuid, x = s.w_output
+
+        assert 1 <= rand <= 10
+        assert len(str(uuid)) == 36
+
+
+@pytest.fixture
+def account_balance():
+    return {"account": 1000}
+
+
+@async_test
+async def test_banking_compensation(t: WorkflowTest, account_balance):
+    def risky_transaction():
+        raise Exception("Transaction failed")
+
+    @workflow()
+    async def banking_workflow(c: Context):
+        def credit_account(value: int) -> int:
+            return account_balance["account"] - value
+
+        def debit_account(value: int) -> int:
+            return account_balance["account"] + value
+
+        c.save(debit_account(4), compensate_with=credit_account(4))
+        c.save(debit_account(10), compensate_with=credit_account(10))
+        c.save(
```

Comment:
Suggestion: add a sleep in between these to force trigger a replay better testing.